src/core/surface/init.c

```diff
@@ -58,7 +61,42 @@ static void do_basic_init(void) {
   g_initializations = 0;
 }
 
+typedef struct grpc_plugin {
+  void (*init)();
+  void (*deinit)();
+  struct grpc_plugin *next;
+} grpc_plugin;
+
+static grpc_plugin *g_plugins_head = NULL;
+
+static grpc_plugin *new_plugin(void (*init)(void), void (*deinit)(void)) {
+  grpc_plugin *plugin = gpr_malloc(sizeof(*plugin));
+  memset(plugin, 0, sizeof(*plugin));
+  plugin->init = init;
+  plugin->deinit = deinit;
+
+  return plugin;
+}
+
+void grpc_register_plugin(void (*init)(void), void (*deinit)(void)) {
+  grpc_plugin *old_head = g_plugins_head;
+  g_plugins_head = new_plugin(init, deinit);
+  g_plugins_head->next = old_head;
+}
+
+void grpc_unregister_all_plugins() {
```

Comment:
Let's actually pull that, and use a static array for the plugins instead of allocating them. 128 entries should be more than enough without wasting memory.